File: src-tauri/src/domain/memory.rs
//! 内存分配算法（对应前端 `src/domain/memory.ts`，ADR-029）
//!
//! 核心思想：**先按"游戏需要多少"定锚，再按"机器还剩多少"打折**。
//! 比"物理内存的一半"准得多，天然适配整合包。
//!
//! ★ 修掉了原设计稿实现里的一个 bug：原稿在"可用内存不足 T1"时仍给足 T1，
//!   导致分配值可能超过可用内存。这里保证**任何阶段都不得超发**。

use serde::Serialize;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InstanceMemoryType {
    Modded,
    OptiFine,
    Vanilla,
}

#[derive(Debug, Clone, Copy, Serialize)]
pub struct MemoryTargets {
    pub min: f64,
    pub t1: f64,
    pub t2: f64,
    pub t3: f64,
}

pub fn memory_targets(mod_count: u32, kind: InstanceMemoryType) -> MemoryTargets {
    let n = mod_count as f64;
    match kind {
        InstanceMemoryType::Modded => MemoryTargets {
            min: 0.5 + n / 150.0,
            t1: 1.5 + n / 90.0,
            t2: 2.7 + n / 50.0,
            t3: 4.5 + n / 25.0,
        },
        InstanceMemoryType::OptiFine => MemoryTargets {
            min: 0.5,
            t1: 1.5,
            t2: 3.0,
            t3: 5.0,
        },
        InstanceMemoryType::Vanilla => MemoryTargets {
            min: 0.5,
            t1: 1.5,
            t2: 2.5,
            t3: 4.0,
        },
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct AutoMemoryResult {
    pub gb: f64,
    pub mod_count: u32,
    pub available_gb: f64,
    pub start_gb: f64,
    pub escalate_gb: f64,
    pub capped_by_available: bool,
}

fn round1(n: f64) -> f64 {
    (n * 10.0).round() / 10.0
}

/// 四阶段递减比例分配：
/// 阶段一 0→T1 取 100%；阶段二 T1→T2 取 70%；阶段三 T2→T3 取 40%；阶段四 T3→2·T3 取 15%。
/// 任一阶段后可用 < 0.1 GB 就停；最后 max(值, 最低值)，并受物理内存封顶。
pub fn auto_memory(
    mod_count: u32,
    kind: InstanceMemoryType,
    total_gb: f64,
    available_gb: f64,
) -> AutoMemoryResult {
    let t = memory_targets(mod_count, kind);
    let mut avail = available_gb.max(0.0);
    let mut give = 0.0;

    // 阶段一：0 → T1，全给
    let s1 = avail.min(t.t1);
    give += s1;
    avail -= s1;
    let capped = s1 < t.t1 - 1e-9;

    if !capped && avail >= 0.1 {
        let s2 = (avail * 0.7).min(t.t2 - t.t1);
        give += s2;
        avail -= s2;
    }
    if give >= t.t2 - 1e-9 && avail >= 0.1 {
        let s3 = (avail * 0.4).min(t.t3 - t.t2);
        give += s3;
        avail -= s3;
    }
    if give >= t.t3 - 1e-9 && avail >= 0.1 {
        let s4 = (avail * 0.15).min(t.t3);
        give += s4;
    }

    let gb = round1(give.max(t.min).min(total_gb.max(0.5)));

    AutoMemoryResult {
        gb,
        mod_count,
        available_gb: round1(available_gb),
        start_gb: round1(1.5 + mod_count as f64 / 90.0),
        escalate_gb: round1(2.7 + mod_count as f64 / 50.0),
        capped_by_available: capped || give < t.t2 - 1e-9,
    }
}

/// 生成"依据"文案 —— 用**真实算出来的数值**，不写死。
/// 原设计稿写死了"向 2.7 GB 递进"，而实际算出的是 3.2，说明文字与结果对不上。
pub fn memory_reasoning(mod_count: u32, r: &AutoMemoryResult) -> String {
    if r.capped_by_available {
        format!(
            "检测到 {mod_count} 个 Mod，按「1.5 + {mod_count}/90 ≈ {} GB」起步；受当前可用内存 {} GB 限制，分配 {} GB。",
            r.start_gb, r.available_gb, r.gb
        )
    } else {
        format!(
            "检测到 {mod_count} 个 Mod，按「1.5 + {mod_count}/90 ≈ {} GB」起步，可用内存充裕时向 {} GB 递进；本次分配 {} GB（可用 {} GB）。",
            r.start_gb, r.escalate_gb, r.gb, r.available_gb
        )
    }
}

/* ====================== 滑块档位映射 ====================== */

/// 档位 → GB 的分段映射（与原版一致，已数值验证）
pub fn gear_to_gb(v: u32) -> f64 {
    if v <= 12 {
        round1(v as f64 * 0.1 + 0.3)
    } else if v <= 25 {
        round1((v as f64 - 12.0) * 0.5 + 1.5)
    } else if v <= 33 {
        round1(v as f64 - 25.0 + 8.0)
    } else {
        round1((v as f64 - 33.0) * 2.0 + 16.0)
    }
}

/// 档位上限随物理内存动态变化
pub fn max_gear(total_gb: f64) -> u32 {
    let g = if total_gb <= 1.5 {
        ((total_gb - 0.3) / 0.1).floor().max(1.0)
    } else if total_gb <= 8.0 {
        ((total_gb - 1.5) / 0.5).floor() + 12.0
    } else if total_gb <= 16.0 {
        (total_gb - 8.0).floor() + 25.0
    } else {
        ((total_gb - 16.0) / 2.0).floor() + 33.0
    };
    g.max(0.0) as u32
}

/// GB → 最接近的档位。
/// ★ 修掉了原稿"手柄位置与显示值对不上"的问题：取**最接近**而不是向下取整，
///   并把该档对应的 GB 作为唯一真值。
pub fn gb_to_gear(gb: f64, total_gb: f64) -> u32 {
    let max = max_gear(total_gb);
    let mut best = 0u32;
    let mut best_diff = f64::INFINITY;
    for v in 0..=max {
        let diff = (gear_to_gb(v) - gb).abs();
        if diff < best_diff {
            best_diff = diff;
            best = v;
        }
    }
    best
}

/// 把任意内存值吸附到滑块能表达的最近档位
pub fn snap_to_gear(gb: f64, total_gb: f64) -> f64 {
    gear_to_gb(gb_to_gear(gb, total_gb))
}

#[derive(Debug, Clone, Serialize)]
pub struct MemoryBar {
    pub used_gb: f64,
    pub game_gb: f64,
    pub free_gb: f64,
    pub total_gb: f64,
    pub over_available: bool,
}

pub fn memory_bar(requested_gb: f64, total_gb: f64, available_gb: f64) -> MemoryBar {
    let used = (total_gb - available_gb).max(0.0);
    let game = requested_gb.min(available_gb);
    let free = (total_gb - used - game).max(0.0);
    MemoryBar {
        used_gb: round1(used),
        game_gb: round1(game),
        free_gb: round1(free),
        total_gb: round1(total_gb),
        over_available: requested_gb > available_gb + 1e-9,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn gear_mapping_matches_original() {
        assert_eq!(gear_to_gb(12), 1.5);
        assert_eq!(gear_to_gb(25), 8.0);
        assert_eq!(gear_to_gb(33), 16.0);
    }

    #[test]
    fn gear_mapping_is_monotonic() {
        let mut prev = -1.0;
        for v in 0..=max_gear(32.0) {
            let gb = gear_to_gb(v);
            assert!(gb >= prev - 1e-9, "档 {v} 回退: {gb} < {prev}");
            prev = gb;
        }
    }

    #[test]
    fn auto_never_overshoots_available() {
        // ★ 回归测试：原设计稿在这里会超发
        let r = auto_memory(24, InstanceMemoryType::Modded, 8.0, 1.5);
        assert!(r.gb <= 1.5 + 1e-9, "分配 {} 超过可用 1.5", r.gb);
        assert!(r.capped_by_available);
    }

    #[test]
    fn auto_grows_with_mod_count() {
        let few = auto_memory(0, InstanceMemoryType::Modded, 32.0, 16.0).gb;
        let many = auto_memory(200, InstanceMemoryType::Modded, 32.0, 16.0).gb;
        assert!(many > few);
    }

    #[test]
    fn snap_keeps_handle_and_label_consistent() {
        let raw = auto_memory(24, InstanceMemoryType::Modded, 7.9, 3.2).gb;
        let g = gb_to_gear(raw, 7.9);
        assert_eq!(gear_to_gb(g), snap_to_gear(raw, 7.9));
    }

    #[test]
    fn bar_segments_add_up() {
        let b = memory_bar(2.5, 7.9, 3.2);
        let sum = b.used_gb + b.game_gb + b.free_gb;
        assert!((sum - b.total_gb).abs() < 0.11, "{sum} vs {}", b.total_gb);
    }
}

File: src/domain/memory.ts
/**
 * 内存分配算法（ADR-029）
 * ------------------------------------------------------------------
 * 核心思想：**先按"游戏需要多少"定锚，再按"机器还剩多少"打折**。
 * 比"物理内存的一半"准得多，天然适配整合包（300 Mod 时 T1 就已经 4.8 GB）。
 *
 * 算法来自 PCL2 `PageInstanceSetup` 的源码研读（源码研读 13.3），
 * 但修掉了原稿实现里的一个 bug：原稿在"可用内存不足 T1"时仍给足 T1，
 * 导致分配值可能超过可用内存。这里改为**任何阶段都不得超发**。
 */

export type InstanceMemoryType = 'modded' | 'optifine' | 'vanilla';

export interface MemoryTargets {
  /** 最低可玩值 */
  min: number;
  /** 一/二/三阶段的上界 */
  t1: number;
  t2: number;
  t3: number;
}

/** 按 Mod 数量与实例类型算四个目标值（GB） */
export function memoryTargets(modCount: number, type: InstanceMemoryType): MemoryTargets {
  switch (type) {
    case 'modded':
      return {
        min: 0.5 + modCount / 150,
        t1: 1.5 + modCount / 90,
        t2: 2.7 + modCount / 50,
        t3: 4.5 + modCount / 25,
      };
    case 'optifine':
      return { min: 0.5, t1: 1.5, t2: 3, t3: 5 };
    case 'vanilla':
    default:
      return { min: 0.5, t1: 1.5, t2: 2.5, t3: 4 };
  }
}

export interface AutoMemoryResult {
  /** 建议分配的 GB —— **整数**（用户："给也是直接给整数内存"） */
  gb: number;
  /** 依据说明里要用的中间值 */
  detail: {
    modCount: number;
    availableGb: number;
    startGb: number;
    /** 递进目标 */
    escalateGb: number;
    /** 被可用内存封顶 */
    cappedByAvailable: boolean;
  };
}

const round1 = (n: number) => Math.round(n * 10) / 10;

/**
 * MB → **整数** GB（只用于显示）。
 *
 * ★ 用户原话（截图里那一行显示 `内存 3.599609375 GB`）：
 *   「内存取整数即可，给也是直接给整数内存」。
 *
 * 为什么必须有个专门的函数，而不是随手写 `Math.round(mb / 1024)`：
 *   ① 存储单位是 MB（`Instance.config.memoryMb`），引擎也算 MB —— 这里**不改语义**，
 *      只把"给人看的那一眼"收敛成整数；
 *   ② `3.599609375` 的直接来源就是 `3686 / 1024`：自动算法给出 3.6 GB 后按
 *      `Math.round(3.6 * 1024)` 存成 3686 MB，再除回来就不再是 3.6 了。
 *      1024 不是 10 的幂，**任何"非整 GB 的 MB 值"除回来都会带一长串小数**，
 *      所以显示层必须取整，否则换个数字还会再犯。
 */
export function wholeGbFromMb(memoryMb: number): number {
  return Math.round(memoryMb / 1024);
}

/**
 * GB → **整数** GB：用户可设置的值（滑块 / 档位）也一律是整数。
 * 同样是用户那句「给也是直接给整数内存」的落点。
 *
 * ★ 这里用 `Math.floor` 而不是 `Math.round`：`autoMemory` 的结果带有
 *   "**绝不超发**"这条硬约束（见下方四阶段说明与它的回归测试），
 *   四舍五入会把 3.6 变成 4，凭空多要 0.4 GB —— 向下取整才守得住这条约束。
 */
export function wholeGb(gb: number): number {
  return Math.floor(gb);
}

/**
 * 四阶段递减比例分配。
 * 阶段一 0→T1 取 100%；阶段二 T1→T2 取 70%；阶段三 T2→T3 取 40%；阶段四 T3→2·T3 取 15%。
 * 任一阶段后可用 < 0.1 GB 就停；最后 max(值, 最低值)。
 *
 * ★ 修正：每个阶段实际给出的量是 `min(本阶段可用额度, 阶段跨度)`，
 *   且**绝不超过剩余可用内存**。
 */
export function autoMemory(
  modCount: number,
  type: InstanceMemoryType,
  totalGb: number,
  availableGb: number,
): AutoMemoryResult {
  const T = memoryTargets(modCount, type);
  let avail = Math.max(availableGb, 0);
  let give = 0;

  // 阶段一：0 → T1，全给
  const s1 = Math.min(avail, T.t1);
  give += s1;
  avail -= s1;
  const capped = s1 < T.t1 - 1e-9;

  if (!capped && avail >= 0.1) {
    // 阶段二：T1 → T2，按 70% 折算
    const s2 = Math.min(avail * 0.7, T.t2 - T.t1);
    give += s2;
    avail -= s2;
  }
  const reachedT2 = give >= T.t2 - 1e-9;
  if (reachedT2 && avail >= 0.1) {
    // 阶段三：T2 → T3，按 40% 折算
    const s3 = Math.min(avail * 0.4, T.t3 - T.t2);
    give += s3;
    avail -= s3;
  }
  const reachedT3 = give >= T.t3 - 1e-9;
  if (reachedT3 && avail >= 0.1) {
    // 阶段四：T3 → 2·T3，按 15% 折算
    const s4 = Math.min(avail * 0.15, T.t3);
    give += s4;
    avail -= s4;
  }

  /*
   * ★ 2026-09-16 用户（截图 `内存 3.599609375 GB`）：
   *   「内存取整数即可，给也是直接给整数内存」。
   *
   *   这里在**算法出口**就取整，而不是只让界面显示整数 —— 两处都要，缺一不可：
   *     · 只改显示：`3.6` 存成 3686 MB、界面写 4 GB，说的和存的不是一回事，
   *       下次再有人从 MB 反推显示，长小数又回来了；
   *     · 只改这里：老实例库里已经有 3686 MB 这类值，界面照样漏出长小数。
   *   所以出口取整（保证新值）＋显示取整（兜住老值，见 `wholeGbFromMb`）。
   *
   *   用 `wholeGb`（向下取整）不是 `Math.round`：上面的四阶段分配有一条
   *   "任何阶段都不得超发"的硬约束（`autoMemory(24,'modded',8,1.5)` 不许超过 1.5），
   *   而 `Math.round(1.5) = 2` 正好会把它破掉。
   *   算法本身一个数都没动，取整只作用在它已经算完的输出上。
   */
  const gb = wholeGb(Math.min(Math.max(give, T.min), Math.max(totalGb, 0.5)));

  return {
    gb,
    detail: {
      modCount,
      availableGb: round1(availableGb),
      startGb: round1(1.5 + modCount / 90),
      escalateGb: round1(2.7 + modCount / 50),
      cappedByAvailable: capped || give < T.t2 - 1e-9,
    },
  };
}

/**
 * 生成"依据"文案。
 * ★ 这里刻意用真实计算出的数值，不写死 —— 原设计稿写死了"向 2.7 GB 递进"，
 *   而实际算出的是 3.24，说明文字与结果对不上。
 */
export function memoryReasoning(modCount: number, r: AutoMemoryResult): string {
  const { startGb, escalateGb, availableGb, cappedByAvailable } = r.detail;
  if (cappedByAvailable) {
    return `检测到 ${modCount} 个 Mod，按「1.5 + ${modCount}/90 ≈ ${startGb} GB」起步；` +
      `受当前可用内存 ${availableGb} GB 限制，分配 ${r.gb} GB。`;
  }
  return `检测到 ${modCount} 个 Mod，按「1.5 + ${modCount}/90 ≈ ${startGb} GB」起步，` +
    `可用内存充裕时向 ${escalateGb} GB 递进；本次分配 ${r.gb} GB（可用 ${availableGb} GB）。`;
}

/* ====================== 滑块档位映射 ====================== */

/**
 * 档位 → GB 的分段映射（与原版一致，已数值验证）
 *   0..12  → 0.1v + 0.3      (0.3 ~ 1.5)
 *   13..25 → 0.5(v-12) + 1.5 (2.0 ~ 8.0)
 *   26..33 → (v-25) + 8      (9 ~ 16)
 *   34..   → 2(v-33) + 16    (18+)
 */
export function gearToGb(v: number): number {
  if (v <= 12) return round1(v * 0.1 + 0.3);
  if (v <= 25) return round1((v - 12) * 0.5 + 1.5);
  if (v <= 33) return round1(v - 25 + 8);
  return round1((v - 33) * 2 + 16);
}

/** 档位上限随物理内存动态变化（避免滑到超过本机内存的值） */
export function maxGear(totalGb: number): number {
  if (totalGb <= 1.5) return Math.max(Math.floor((totalGb - 0.3) / 0.1), 1);
  if (totalGb <= 8) return Math.floor((totalGb - 1.5) / 0.5) + 12;
  if (totalGb <= 16) return Math.floor((totalGb - 8) / 1) + 25;
  return Math.floor((totalGb - 16) / 2) + 33;
}

/**
 * GB → 最接近的档位（滑块定位用）
 * ★ 修掉了原稿的"档位与显示值不一致"：原稿先把滑块设到某档，再显示自动值，
 *   于是手柄位置（1.7 GB）和数字（2.5 GB）对不上。这里取**最接近**的档位，
 *   并把该档对应的 GB 作为唯一真值返回，保证手柄与数字永远一致。
 */
export function gbToGear(gb: number, totalGb: number): number {
  const max = maxGear(totalGb);
  let best = 0;
  let bestDiff = Infinity;
  for (let v = 0; v <= max; v++) {
    const diff = Math.abs(gearToGb(v) - gb);
    if (diff < bestDiff) {
      bestDiff = diff;
      best = v;
    }
  }
  return best;
}

/**
 * 把任意内存值吸附到滑块能表达的最近档位，返回吸附后的值。
 * UI 应使用这个函数，而不是直接用自动算法的原始值。
 */
export function snapToGear(gb: number, totalGb: number): number {
  return gearToGb(gbToGear(gb, totalGb));
}

/** 内存占用三分段的展示数据 */
export interface MemoryBar {
  usedGb: number;
  gameGb: number;
  freeGb: number;
  totalGb: number;
  /** 请求值是否超过可用内存 */
  overAvailable: boolean;
}

export function memoryBar(
  requestedGb: number,
  totalGb: number,
  availableGb: number,
): MemoryBar {
  const used = Math.max(totalGb - availableGb, 0);
  const game = Math.min(requestedGb, availableGb);
  const free = Math.max(totalGb - used - game, 0);
  return {
    usedGb: round1(used),
    gameGb: round1(game),
    freeGb: round1(free),
    totalGb: round1(totalGb),
    overAvailable: requestedGb > availableGb + 1e-9,
  };
}
